src/librbd/cache/pwl/rwl/WriteLog.cc

```diff
@@ -475,7 +475,7 @@ bool WriteLog<I>::retire_entries(const unsigned long int frees_per_tx) {
     std::lock_guard locker(m_lock);
     initial_first_valid_entry = this->m_first_valid_entry;
     first_valid_entry = this->m_first_valid_entry;
-    auto entry = m_log_entries.front();
+    auto entry = m_log_entries.empty() ? nullptr : m_log_entries.front();
     while (!m_log_entries.empty() &&
            retiring_entries.size() < frees_per_tx &&
            this->can_retire_entry(entry)) {
```

Comment:
nit: For me a little cleaner approach seems to use `this->can_retire_entry(m_log_entries.front())` here, and move `auto entry = m_log_entries.front();` inside this `while` block as the first statement (and remove `entry = m_log_entries.front();` at the end of the block). What do you think?

Reply:
Good suggestion, thanks.  At the same time, it prevents getting the empty `m_log_entries.front() ` at the end. 
I have an idea to explore further. if `auto entry ` is in the `while` block, it should allocate for registers and release them every loop. Would it be better to put the declaration outside the `while` block. For example, I declare on `while` without assignment. Is this would be better and be allowed?
```
std::shared_ptr<GenericLogEntry> entry;
while {}
```

Reply:
Personally, I don't like optimization like this unless it gives a real gain, and I have no idea (or rather doubt) if it does in this case. Actually, I would expect some optimization from the compiler here that would make programmer not bother about things like this.

And it would make the code look a bit confusing: after this change, when looking at the code my first thought would be that the `entry` was going to be used outside the loop and I would spend some time looking at the code and trying to understand where it was.

Anyway, I don't have a strong opinion about this and would be fine with any final version.

Reply:
Agree. The compiler may have optimized, and this makes the code harder to read. Then keep it that way.